src/main/java/org/kairosdb/datastore/cassandra/DataCache.java

```diff
@@ -68,49 +66,63 @@ public DataCache(int cacheSize)
 	}
 
 	/**
-	 returns null if item was not in cache.  If the return is not null the item
+	 returns null if item is not in cache.  If the return is not null the item
 	 from the cache is returned.
 
 	 @param cacheData
 	 @return
 	 */
-	public T cacheItem(T cacheData)
+	public synchronized T get(T cacheData)
 	{
-		LinkItem<T> mappedItem = null;
+		final LinkItem<T> mappedItem = getMappedItem(cacheData);
+
+		pruneCache();
+
+		return (mappedItem == null ? null : mappedItem.m_data);
+	}
 
-		synchronized (m_lock)
+	private synchronized LinkItem<T> getMappedItem(T cacheData) {
+		final LinkItem<T> mappedItem = m_hashMap.get(cacheData);
+
+		if (mappedItem != null)
 		{
-			LinkItem<T> li = new LinkItem<T>(cacheData);
-			mappedItem = m_hashMap.putIfAbsent(cacheData, li);
-
-			if (mappedItem != null)
-			{
-				//moves item to top of list
-				remove(mappedItem);
-				addItem(mappedItem);
-			}
-			else
-				addItem(li);
-
-			if (m_hashMap.size() > m_maxSize)
-			{
-				LinkItem<T> last = m_back.m_prev;
-				remove(last);
-
-				m_hashMap.remove(last.m_data);
-			}
+			//moves item to top of list
+			removeLRUItem(mappedItem);
+			addLRUItem(mappedItem);
 		}
 
-		return (mappedItem == null ? null : mappedItem.m_data);
+		return mappedItem;
+	}
+
+
+	public synchronized void put(final T cacheData) {
+		final LinkItem<T> existing = getMappedItem(cacheData);
```

Comment:
The call to `getMappedItem` performs an LRU update. So under the current cache access pattern, an item's LRU is updated on get and put, which are called for all items even if the get resulted in a cache hit. Other than some wasted cycles, the only impact I see is that under failures requests which fail to write, and thus don't perform the put will be noted as less frequently used (when that's not really the case). Is there a reason we could not just do a hash map get here?

Reply:
If you're trying to `put` an item the second time, that's probably a clue that it'll get used?

But your suggestion shouldn't affect the correctness, so I've simply done it your way.